source/grammar/qasm3.g4

```diff
@@ -256,63 +256,132 @@ quantumGateName
     | quantumGateModifier quantumGateName
     ;
 
-/* Classical Instructions */
+/*** Classical Instructions ***/
 
 unaryOperator
     : '~' | '!'
     ;
 
-binaryOperator
-    : '+' | '-' | '*' | '/' | '<<' | '>>' | 'rotl' | 'rotr' | '&&' | '||' | '&' | '|' | '^'
-    | '>' | '<' | '>=' | '<=' | '==' | '!='
+relationalOperator
+    : '>'
+    | '<'
+    | '>='
+    | '<='
+    | '=='
+    | '!='
+    ;
+
+logicalOperator
+    : '&&'
+    | '||'
     ;
 
 expressionStatement
     : expression SEMICOLON
     ;
 
 expression
-    : expression binaryOperator expression
-    | unaryOperator expression
-    | expression incrementor
-    | expression LBRACKET expression RBRACKET
-    | LPAREN expression RPAREN
-    | membershipTest
-    | builtInCall
-    | subroutineCall
-    | kernelCall
-    | MINUS expression
-    | expressionTerminator
+    // include terminator/unary as base cases to simplify parsing
+    : expressionTerminator
+    | unaryExpression
+    // expression hierarchy
+    | xOrExpression
+    | expression '|' xOrExpression
+    ;
+
+/**  Expression hierarchy for non-terminators. Adapted from ANTLR4 C
+  *  grammar: https://github.com/antlr/grammars-v4/blob/master/c/C.g4
+  * Order (first to last evaluation):
+    Terminator (including Parens),
+    Unary Op,
+    Multiplicative
+    Additive
+    Bit Shift
+    Bit And
+    Exlusive Or (xOr)
+    Bit Or
+**/
+xOrExpression
+    : bitAndExpression
+    | xOrExpression '^' bitAndExpression
+    ;
+
+bitAndExpression
+    : bitShiftExpression
+    | bitAndExpression '&' bitShiftExpression
+    ;
+
+bitShiftExpression
+    : additiveExpression
+    | bitShiftExpression ( '<<' | '>>' ) additiveExpression
+    ;
+
+additiveExpression
+    : multiplicativeExpression
+    | additiveExpression ( '+' | MINUS ) multiplicativeExpression
```

Comment:
Why is `MINUS` special?

Reply:
Addressed in https://github.com/Qiskit/openqasm/pull/157/commits/11c2f62f2dc6420d443a6a5f966c8b198527079c.